### PR TITLE
Added usable prefix field for oLj command

### PR DESCRIPTION
**Pull request by deepakchethan**
For issue #12620

Example output for `oLj` now is:

`{"IO_Plugins":[{"Permissions":"rw_","Name":"ar","Description":"Open ar/lib files [ar|lib]://[file//path]","License":"LGPL3","prefix":"ar://, lib://"},{"Permissions":"rw_","Name":"bfdbg","Description":"BrainFuck Debugger (bfdbg://path/to/file)","License":"LGPL3","prefix":"bfdbg://"}....}`

Please let me know if any more changes are to be done

### CODE REVIEW

File: libr/io/plugin.c [hunk 1]
```diff
@@ -158,3 +162,105 @@ R_API int r_io_plugin_write_at(RIODesc *desc, ut64 addr, const ut8 *buf, int len
 	}
 	return 0;
 }
+
+R_API char *r_io_plugin_get_uri_prefix (RIOPlugin *plugin) {
```
Comment (radare) [inline libr/io/plugin.c]: wtf? why is this necessary? why is this hardcoded?

File: libr/io/plugin.c [hunk 2]
```diff
@@ -158,3 +162,105 @@ R_API int r_io_plugin_write_at(RIODesc *desc, ut64 addr, const ut8 *buf, int len
...
+	char *prefix = NULL;
+	if (!strcmp ("ar", pluginName)) {
+		prefix = "ar://, lib://";
+	}
+	else if (!strcmp ("bfdbg", pluginName)) {
+		prefix = "bfdbg://";
+	}
+	else if (!strcmp ("bfdbg", pluginName)) {
+		prefix = "bfdbg://";
+	}
```
Comment (radare) [inline libr/io/plugin.c]: else goes in the same line as the }

File: libr/io/plugin.c (hunk 1 above)
Comment (radare) [reply libr/io/plugin.c]: return type should be const char *

**Comment by radare** [comment]
this information must be provided by the plugin, not hardcoded in any place

File: libr/io/plugin.c (hunk 1 above)
Comment (deepakchethan) [reply libr/io/plugin.c]: All the plugin prefixes are hardcoded in r_io_ar_plugin_open() method like
```
static bool r_io_ar_plugin_open(RIO *io, const char *file, bool many) {
	return !strncmp ("ar://", file, 5) || !strncmp ("lib://", file, 6);
}
````
and there are no internal fields that specify the prefix
I made another pr by adding an internal field inside the RIOPlugin and modifying 34 files. Since you told me not to modify internal fields a couple of times. I closed that PR and changed to this.

The PR in question is #13012. 

Thanks for taking the time to review. If both my ways are wrong. Feel free to close this issue with a message pointing the right direction.

**Comment by deepakchethan** [comment]
@radare thanks for nudging me in the right direction.

Since there was no internal field in RIOPlugin about the list of prefixes, I have created one. Initialized the same on all available plugins.

Could you review the code again ?

File: libr/io/p/io_ar.c [hunk 3]
```diff
@@ -94,6 +94,7 @@ RIOPlugin r_io_plugin_ar = {
 	.name = "ar",
 	.desc = "Open ar/lib files [ar|lib]://[file//path]",
 	.license = "LGPL3",
+	.prefixes = "ar://, lib://",
```
Comment (radare) [inline libr/io/p/io_ar.c]: what about making this field an actual array of char* instead of a plain string?

File: libr/include/r_io.h [hunk 4]
```diff
@@ -158,6 +158,7 @@ typedef struct r_io_plugin_t {
 	char *author;
 	char *license;
 	void *widget;
+	char *prefixes;
```
Comment (radare) [inline libr/include/r_io.h]: thats wrong, this buffer cant be modified, use const char *.. or const char ** to use {"foo","bar",NULL}

**Comment by radare** [comment]
also the :// is called URI handler, not "prefix", just to clarify concepts and find a better name for that field

**Comment by deepakchethan** [comment]
> also the :// is called URI handler, not "prefix", just to clarify concepts and find a better name for that field

Okay @radare, thanks for letting me know. So I'll change it to `uri_handlers` then?

File: libr/include/r_io.h (hunk 4 above)
Comment (deepakchethan) [reply libr/include/r_io.h]: I will change it to a const buffer.

File: libr/io/p/io_ar.c (hunk 3 above)
Comment (deepakchethan) [reply libr/io/p/io_ar.c]: char ** type cannot be directly assigned like `char **a = {"one", "two", NULL}`. One way is using `char name[]` but even that requires using `strcpy` if we are assigning after declaration. 
Also since it has no other use other than directly displaying, is it really required?

File: libr/io/p/io_ar.c [hunk 5]
```diff
@@ -94,6 +94,7 @@ RIOPlugin r_io_plugin_ar = {
 	.name = "ar",
 	.desc = "Open ar/lib files [ar|lib]://[file//path]",
 	.license = "LGPL3",
+	.uri_handlers = "ar://, lib://",
```
Comment (xarkes) [inline libr/io/p/io_ar.c]: Why don't you use an actual list rather than a plain `char *`?

File: libr/io/p/io_ar.c (hunk 3 above)
Comment (xarkes) [reply libr/io/p/io_ar.c]: Sorry I didn't see that comment, ideally the json should yield a list for the available prefixes, so it's actually parsable :)
Something like:
```
{"IO_Plugins":[{"Permissions":"rw_","Name":"ar","Description":"Open ar/lib files [ar|lib]://[file//path]","License":"LGPL3","prefixex": ["ar://", "lib://"]}....}
```

Comment (deepakchethan) [reply libr/io/p/io_ar.c]: Okay xarxes, I will make the change and push them by tomorrow.

**Comment by radare** [comment]
Const char* and it should work if u define a reference

**Comment by radare** [comment]
I prefer just “uris”

File: libr/include/r_io.h [hunk 6]
```diff
@@ -158,6 +158,7 @@ typedef struct r_io_plugin_t {
 	char *author;
 	char *license;
 	void *widget;
+	const char *uri_handlers;
```
Comment (radare) [inline libr/include/r_io.h]: rename to uris

File: libr/io/p/io_ar.c (hunk 5 above)
Comment (radare) [reply libr/io/p/io_ar.c]: const char ** should be fine, but we can also use comma separated values and use r_str_split_list, but then i would not add a space after the comma

**Comment by radare** [comment]
btw for simplicity creating bindings or plugins in other languages (rust, vala, python ,...) having just char * instead of char ** would be easier, so maybe we can just go for the comma separated string, without a space

**Comment by deepakchethan** [comment]
Okay @radare, I will make the changes

**Comment by deepakchethan** [comment]
I have converted the field to `uris` and removed the spaces in the `uris` field on all plugins. In the json parsing code. I used the split method to split the uris and display them in the form of the array. 

The raw output now is:
`{"IO_Plugins":[{"permissions":"rw_","name":"ar","description":"Open ar/lib files [ar|lib]://[file//path]","license":"LGPL3","uris":["ar://","lib://"]},{"permissions":"rw_","name":"bfdbg","description":"BrainFuck Debugger (bfdbg://path/to/file)","license":"LGPL3","uris":["bfdbg://"]}..`


The parsed output is: 
screen shot 2019-02-10 at 10 54 50 pm

Ready for the review @radare

**Comment by radare** [comment]
now we can probably remove the uris from the description, and show them in `r2 -L`

File: libr/io/plugin.c [hunk 7]
```diff
@@ -100,28 +100,53 @@ R_API int r_io_plugin_list(RIO *io) {
...
+			char *uri;
+			char *uris = strdup (plugin->uris);
+			RList *plist = r_str_split_list (uris, ",");
+			RListIter *piter;
+			pj_k (pj, "uris");
+			pj_a (pj);
+			r_list_foreach (plist, piter, uri) {
+				pj_s (pj, uri);
+			}
+			pj_end (pj);
```
Comment (radare) [inline libr/io/plugin.c]: r_list_free (plist); in here

Comment (radare) [reply libr/io/plugin.c]: free (uris); its leaking here too

File: libr/io/plugin.c [hunk 8]
```diff
@@ -100,28 +100,53 @@ R_API int r_io_plugin_list(RIO *io) {
...
+			pj_ks (pj, "author", plugin->author);
 		}
-		io->cb_printf ("}");
+		pj_end (pj);
 		n++;
 	}
-	io->cb_printf("]}");
+	pj_end (pj);
+	pj_end (pj);
+	io->cb_printf (pj_string(pj));
```
Comment (radare) [inline libr/io/plugin.c]: missing space before (

File: libr/io/plugin.c [hunk 9]
```diff
@@ -100,28 +100,53 @@ R_API int r_io_plugin_list(RIO *io) {
 R_API int r_io_plugin_list_json(RIO *io) {
 	RIOPlugin *plugin;
 	SdbListIter *iter;
+	PJ *pj = pj_new();
```
Comment (radare) [inline libr/io/plugin.c]: missing space

File: libr/io/p/io_zip.c [hunk 10]
```diff
@@ -651,6 +651,7 @@ static int r_io_zip_close(RIODesc *fd) {
 RIOPlugin r_io_plugin_zip = {
 	.name = "zip",
 	.desc = "Open zip files [apk|ipa|zip|zipall]://[file//path]",
```
Comment (radare) [inline libr/io/p/io_zip.c]: remove uri information in the description and improve the text

File: libr/io/p/io_http.c [hunk 11]
```diff
@@ -104,7 +104,7 @@ static RIODesc *__open(RIO *io, const char *pathname, int rw, int mode) {
 
 RIOPlugin r_io_plugin_http = {
 	.name = "http",
-	.desc = "http get (http://rada.re/)",
+	.desc = "Make http get requests",
```
[the remark this answers is not in the document]
Comment (deepakchethan) [reply libr/io/p/io_http.c]: Done radare. Anymore changes?

File: libr/io/plugin.c [hunk 12]
```diff
@@ -100,28 +103,55 @@ R_API int r_io_plugin_list(RIO *io) {
...
+	PJ *pj = pj_new ();
+	if (!pj) {
+		return NULL;
+	}
+	
 	char str[4];
 	int n = 0;
-	io->cb_printf("{\"IO_Plugins\":[");
+	pj_o (pj);
+	pj_k (pj, "IO_Plugins");
```
Comment (radare) [inline libr/io/plugin.c]: i dont like this camel snake case. maybe better to just use "io_plugins"

**Comment by radare** [comment]
please rebase

File: libr/io/plugin.c (hunk 12 above)
Comment (deepakchethan) [reply libr/io/plugin.c]: I changed it.

**Comment by deepakchethan** [comment]
I have squashed and rebased the commits @radare 

Thanks for helping me help with this issue.
